test/fgpu/test_send.py

```diff
@@ -0,0 +1,285 @@
+from katgpucbf.utils import TimeConverter, gaussian_dtype
+
+from .. import PromDiff, unpack_complex
+
+pytest_mark = pytest.mark.parametrize("bits", [4, 8])
+N_SUBSTREAMS = 16
+N_CHUNKS = 5
+N_FRAMES = 7  # frames per chunk
+N_CHANNELS = 1024
+N_SPECTRA_PER_HEAP = 32  # Small to make the test fast
+SPECTRA_SAMPLES = 2 * N_CHANNELS
+FENG_ID = 3
+NAME = "foo"
+
+
+@pytest.fixture(params=[4, 8])
+def sample_bits(request) -> int:
+    return request.param
+
+
+@pytest.fixture(params=[1, 2])
+def interfaces(request) -> Sequence[str]:
+    # request.param gives number of interfaces to use
+    return ["10.0.0.1", "10.0.0.2"][: request.param]
+
+
+@pytest.fixture
+def time_converter() -> TimeConverter:
+    return TimeConverter(1234567890.0, 1e9)
+
+
+@pytest.fixture
+def queues(interfaces: Sequence[str]) -> dict[str, list[spead2.InprocQueue]]:
+    """Create in-process queues to connect the test code to the test."""
+    return {iface: [spead2.InprocQueue() for _ in range(N_SUBSTREAMS)] for iface in interfaces}
+
+
+@pytest.fixture
+def chunks(sample_bits) -> list[Chunk]:
+    dtype = gaussian_dtype(sample_bits)
+    return [
+        Chunk(
+            np.zeros((N_FRAMES, N_CHANNELS, N_SPECTRA_PER_HEAP, N_POLS), dtype),
+            np.zeros((N_FRAMES, N_POLS), np.uint32),
+            n_substreams=N_SUBSTREAMS,
+            feng_id=FENG_ID,
+            spectra_samples=SPECTRA_SAMPLES,
+        )
+        for _ in range(N_CHUNKS)
+    ]
+
+
+@pytest.fixture
+def send_streams(
+    queues: dict[str, list[spead2.InprocQueue]], chunks: list[Chunk]
+) -> list["spead2.send.asyncio.AsyncStream"]:
+    """Create the send streams.
+
+    The actual stream constructors are mocked so that we use in-process
+    streams.
+    """
+    expected_endpoints = endpoint_list_parser(None)(f"239.102.1.0+{N_SUBSTREAMS - 1}:7148")
+
+    # Note: this function signature is somewhat fragile. It's built to match the
+    # call in send.py.
+    def make_inproc_stream(
+        thread_pool: spead2.ThreadPool,
+        endpoints: list[tuple[str, int]],
+        config: spead2.send.StreamConfig,
+        interface_address: str,
+        **kwargs,
+    ) -> spead2.send.asyncio.InprocStream:
+        assert endpoints == [tuple(x) for x in expected_endpoints]
+        return spead2.send.asyncio.InprocStream(thread_pool, queues[interface_address], config)
+
+    with mock.patch("spead2.send.asyncio.UdpStream", make_inproc_stream):
+        return make_streams(
+            output_name=NAME,
+            thread_pool=spead2.ThreadPool(1),
+            endpoints=expected_endpoints,
+            interfaces=list(queues.keys()),
+            ttl=4,
+            ibv=False,
+            packet_payload=8192,
+            comp_vector=0,
+            buffer=65536,
+            bandwidth=1e9,
+            send_rate_factor=1.05,
+            feng_id=FENG_ID,
+            num_ants=64,
+            n_data_heaps=N_CHUNKS * N_FRAMES * N_SUBSTREAMS,
+            chunks=chunks,
+        )
```

Comment:
I realise it's probably a bit difficult to declare these naked constants elsewhere. Could I at least request a one/two-liner comment around "these are typical values for each parameter", perhaps even touch on why e.g. `1e9` was chosen for `bandwidth`?